File: app/src/main/java/org/secfirst/umbrella/util/IOCipherContentProvider.java
package org.secfirst.umbrella.util;

import android.content.ContentProvider;
import android.content.ContentValues;
import android.database.Cursor;
import android.database.MatrixCursor;
import android.net.Uri;
import android.os.ParcelFileDescriptor;
import android.provider.OpenableColumns;
import android.util.Log;
import android.webkit.MimeTypeMap;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import info.guardianproject.iocipher.File;
import info.guardianproject.iocipher.FileInputStream;
import info.guardianproject.iocipher.VirtualFileSystem;

import static java.util.Arrays.copyOf;

public class IOCipherContentProvider extends ContentProvider {
    public static final String TAG = "IOCipherContentProvider";
    public static final String SHARE_FILE = "share.db";
    public static final Uri FILES_URI = Uri
            .parse("content://org.secfirst.umbrella/");
    private MimeTypeMap mimeTypeMap;
    private static final String[] COLUMNS = {
            OpenableColumns.DISPLAY_NAME, OpenableColumns.SIZE };

    @Override
    public boolean onCreate() {
        mimeTypeMap = MimeTypeMap.getSingleton();
        return true;
    }

    @Override
    public String getType(Uri uri) {
        String fileExtension = MimeTypeMap.getFileExtensionFromUrl(uri.toString());
        return mimeTypeMap.getMimeTypeFromExtension(fileExtension);
    }

    @Override
    public ParcelFileDescriptor openFile(Uri uri, String mode)
            throws FileNotFoundException {
        ParcelFileDescriptor[] pipe = null;
        InputStream in = null;

        try {
            pipe = ParcelFileDescriptor.createPipe();
            String path = uri.getPath();
            Log.i(TAG, "streaming " + path);
            // BufferedInputStream could help, AutoCloseOutputStream conflicts
            in = new FileInputStream(new File(path));
            new PipeFeederThread(in, new ParcelFileDescriptor.AutoCloseOutputStream(pipe[1])).start();
        } catch (IOException e) {
            Log.e(TAG, "Error opening pipe", e);
            throw new FileNotFoundException("Could not open pipe for: "
                    + uri.toString());
        }

        return (pipe[0]);
    }

    @Override
    public Cursor query(Uri uri, String[] projection, String selection, String[] selectionArgs,
                        String sortOrder) {
        java.io.File f = new java.io.File(Global.INSTANCE.getCacheDir(), SHARE_FILE);
        if (!f.exists()) {
            throw new RuntimeException("Operation not supported");
        }
        String path = f.getPath();

        VirtualFileSystem vfs = VirtualFileSystem.get();
        vfs.setContainerPath(path);
        if (!vfs.isMounted()) vfs.mount(path, Global.INSTANCE.getOrmHelper().getPassword());

        File file = new File("/", uri.getPath().replace(FILES_URI.getPath(),""));

        if (projection == null) {
            projection = COLUMNS;
        }

        String[] cols = new String[projection.length];
        Object[] values = new Object[projection.length];
        int i = 0;
        for (String col : projection) {
            if (OpenableColumns.DISPLAY_NAME.equals(col)) {
                cols[i] = OpenableColumns.DISPLAY_NAME;
                values[i++] = file.getName();
            } else if (OpenableColumns.SIZE.equals(col)) {
                cols[i] = OpenableColumns.SIZE;
                values[i++] = file.length();
            }
        }

        cols = copyOf(cols, i);
        values = copyOf(values, i);

        final MatrixCursor cursor = new MatrixCursor(cols, 1);
        cursor.addRow(values);
        return cursor;
    }

    @Override
    public Uri insert(Uri uri, ContentValues initialValues) {
        throw new RuntimeException("Operation not supported");
    }

    @Override
    public int update(Uri uri, ContentValues values, String where,
                      String[] whereArgs) {
        throw new RuntimeException("Operation not supported");
    }

    @Override
    public int delete(Uri uri, String where, String[] whereArgs) {
        throw new RuntimeException("Operation not supported");
    }

    static class PipeFeederThread extends Thread {
        InputStream in;
        OutputStream out;

        PipeFeederThread(InputStream in, OutputStream out) {
            this.in = in;
            this.out = out;
        }

        @Override
        public void run() {
            byte[] buf = new byte[8192];
            int len;

            try {
                while ((len = in.read(buf)) > 0) {
                    out.write(buf, 0, len);
                }

                in.close();
                out.flush();
                out.close();
            } catch (IOException e) {
                Log.e(TAG, "File transfer failed:", e);
            }
        }
    }
}

File: app/src/main/java/org/secfirst/umbrella/whitelabel/misc/IOCipherContentProvider.kt
package org.secfirst.umbrella.whitelabel.misc

import android.content.ContentProvider
import android.content.ContentValues
import android.database.Cursor
import android.database.MatrixCursor
import android.net.Uri
import android.os.ParcelFileDescriptor
import android.provider.OpenableColumns
import android.system.Os.read
import android.util.Log
import android.webkit.MimeTypeMap

import java.io.FileNotFoundException
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream

import info.guardianproject.iocipher.File
import info.guardianproject.iocipher.FileInputStream
import info.guardianproject.iocipher.VirtualFileSystem
import org.secfirst.umbrella.whitelabel.UmbrellaApplication

import java.util.Arrays.copyOf

class IOCipherContentProvider : ContentProvider() {
    private var mimeTypeMap: MimeTypeMap? = null

    override fun onCreate(): Boolean {
        mimeTypeMap = MimeTypeMap.getSingleton()
        return true
    }

    override fun getType(uri: Uri): String? {
        val fileExtension = MimeTypeMap.getFileExtensionFromUrl(uri.toString())
        return mimeTypeMap!!.getMimeTypeFromExtension(fileExtension)
    }

    @Throws(FileNotFoundException::class)
    override fun openFile(uri: Uri, mode: String): ParcelFileDescriptor? {
        var pipe: Array<ParcelFileDescriptor>? = null
        var `in`: InputStream? = null

        try {
            pipe = ParcelFileDescriptor.createPipe()
            val path = uri.path
            Log.i(TAG, "streaming $path")
            // BufferedInputStream could help, AutoCloseOutputStream conflicts
            `in` = FileInputStream(File(path))
            PipeFeederThread(`in`, ParcelFileDescriptor.AutoCloseOutputStream(pipe!![1])).start()
        } catch (e: IOException) {
            Log.e(TAG, "Error opening pipe", e)
            throw FileNotFoundException("Could not open pipe for: " + uri.toString())
        }

        return pipe[0]
    }

    override fun query(uri: Uri, projection: Array<String>?, selection: String?, selectionArgs: Array<String>?,
                       sortOrder: String?): Cursor? {
        var projection = projection
        val f = java.io.File(UmbrellaApplication.instance.cacheDir, SHARE_FILE)
        if (!f.exists()) {
            throw RuntimeException("Operation not supported")
        }
        val path = f.getPath()

        val vfs = VirtualFileSystem.get()
        vfs.containerPath = path
        if (!vfs.isMounted) vfs.mount(path)

        val file = File("/", uri.path.replace(FILES_URI.path, ""))

        if (projection == null) {
            projection = COLUMNS
        }

        var cols = arrayOfNulls<String>(projection.size)
        var values = arrayOfNulls<Any>(projection.size)
        var i = 0
        for (col in projection) {
            if (OpenableColumns.DISPLAY_NAME == col) {
                cols[i] = OpenableColumns.DISPLAY_NAME
                values[i++] = file.name
            } else if (OpenableColumns.SIZE == col) {
                cols[i] = OpenableColumns.SIZE
                values[i++] = file.length()
            }
        }

        cols = copyOf<String>(cols, i)
        values = copyOf<Any>(values, i)

        val cursor = MatrixCursor(cols, 1)
        cursor.addRow(values)
        return cursor
    }

    override fun insert(uri: Uri, initialValues: ContentValues?): Uri? {
        throw RuntimeException("Operation not supported")
    }

    override fun update(uri: Uri, values: ContentValues?, where: String?,
                        whereArgs: Array<String>?): Int {
        throw RuntimeException("Operation not supported")
    }

    override fun delete(uri: Uri, where: String?, whereArgs: Array<String>?): Int {
        throw RuntimeException("Operation not supported")
    }

    internal class PipeFeederThread(var input: InputStream, private var out: OutputStream) : Thread() {

        override fun run() {
            val buf = ByteArray(8192)
            var len: Int = input.read(buf)

            try {
                while (len > 0) {
                    out.write(buf, 0, len)
                }

                input.close()
                out.flush()
                out.close()
            } catch (e: IOException) {
                Log.e(TAG, "File transfer failed:", e)
            }

        }
    }

    companion object {
        val TAG = "IOCipherContentProvider"
        val SHARE_FILE = "share.db"
        val FILES_URI = Uri
                .parse("content://org.secfirst.umbrella/")
        private val COLUMNS = arrayOf(OpenableColumns.DISPLAY_NAME, OpenableColumns.SIZE)
    }
}